File: android/app/src/main/java/com/crispy/tv/accounts/SupabaseAccountClient.kt
package com.crispy.tv.accounts

import android.content.Context
import com.crispy.tv.network.CrispyHttpClient
import com.crispy.tv.network.CrispyHttpResponse
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.Headers
import okhttp3.HttpUrl.Companion.toHttpUrl
import org.json.JSONArray
import org.json.JSONObject

class SupabaseAccountClient(
    appContext: Context,
    private val httpClient: CrispyHttpClient,
    supabaseUrl: String,
    private val supabasePublishableKey: String,
) {
    private val baseUrl: String = supabaseUrl.trim().trimEnd('/')
    private val prefs = appContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val sessionMutex = Mutex()

    @Volatile
    private var cachedSession: Session? = null

    data class Session(
        val accessToken: String,
        val refreshToken: String,
        val expiresAtEpochSec: Long?,
        val userId: String?,
        val email: String?,
        val anonymous: Boolean,
    )

    data class SignUpResult(
        val session: Session?,
        val message: String,
    )

    fun isConfigured(): Boolean {
        return baseUrl.isNotBlank() && supabasePublishableKey.isNotBlank()
    }

    fun clearLocalSession() {
        saveSession(null)
    }

    fun currentSession(): Session? {
        return loadSession()
    }

    suspend fun ensureValidSession(): Session? {
        val existing = loadSession() ?: return null
        if (!shouldRefresh(existing)) {
            return existing
        }

        return sessionMutex.withLock {
            val latest = loadSession() ?: return@withLock null
            if (!shouldRefresh(latest)) {
                return@withLock latest
            }
            if (latest.refreshToken.isBlank()) {
                saveSession(null)
                return@withLock null
            }

            when (val refreshResult = refreshSession(latest.refreshToken)) {
                is RefreshResult.Success -> {
                    saveSession(refreshResult.session)
                    refreshResult.session
                }

                RefreshResult.InvalidSession -> {
                    saveSession(null)
                    null
                }

                is RefreshResult.TransientFailure -> latest
            }
        }
    }

    suspend fun signInWithEmail(email: String, password: String): Session {
        checkConfigured()
        val url = "$baseUrl/auth/v1/token?grant_type=password".toHttpUrl()
        val payload = JSONObject().put("email", email.trim()).put("password", password).toString()
        val response = httpClient.postJson(url, payload, baseHeaders(), callTimeoutMs = CALL_TIMEOUT_MS)
        val body = requireSuccess(response)
        val session = parseSession(JSONObject(body))
            ?: throw IllegalStateException("Sign-in did not return a session.")
        saveSession(session)
        return session
    }

    suspend fun signUpWithEmail(email: String, password: String): SignUpResult {
        checkConfigured()
        val url = "$baseUrl/auth/v1/signup".toHttpUrl()
        val payload = JSONObject().put("email", email.trim()).put("password", password).toString()
        val response = httpClient.postJson(url, payload, baseHeaders(), callTimeoutMs = CALL_TIMEOUT_MS)
        val body = requireSuccess(response)
        val json = JSONObject(body)
        val session = parseSession(json)
        if (session != null) {
            saveSession(session)
            return SignUpResult(session = session, message = "Account created and signed in.")
        }
        val hasUser = json.optJSONObject("user")?.optString("id")?.isNotBlank() == true
        return if (hasUser) {
            SignUpResult(
                session = null,
                message = "Account created. Confirm your email, then sign in."
            )
        } else {
            SignUpResult(session = null, message = "Account created.")
        }
    }

    suspend fun signOut() {
        if (!isConfigured()) {
            saveSession(null)
            return
        }
        val session = loadSession()
        if (session != null) {
            runCatching {
                val url = "$baseUrl/auth/v1/logout".toHttpUrl()
                httpClient.postJson(url, "{}", authHeaders(session.accessToken), callTimeoutMs = CALL_TIMEOUT_MS)
            }
        }
        saveSession(null)
    }

    private fun checkConfigured() {
        if (!isConfigured()) throw IllegalStateException("Supabase is not configured.")
    }

    private suspend fun refreshSession(refreshToken: String): RefreshResult {
        val url = "$baseUrl/auth/v1/token?grant_type=refresh_token".toHttpUrl()
        val payload = JSONObject().put("refresh_token", refreshToken).toString()
        val response =
            runCatching {
                httpClient.postJson(url, payload, baseHeaders(), callTimeoutMs = CALL_TIMEOUT_MS)
            }.getOrElse { error ->
                return RefreshResult.TransientFailure(error.message)
            }

        if (response.code !in 200..299) {
            return if (isInvalidRefreshResponse(response)) {
                RefreshResult.InvalidSession
            } else {
                RefreshResult.TransientFailure(extractErrorMessage(response.body))
            }
        }

        val session = parseSession(JSONObject(response.body))
            ?: return RefreshResult.TransientFailure("Refresh did not return a session.")
        return RefreshResult.Success(session)
    }

    private fun parseSession(json: JSONObject): Session? {
        val accessToken = json.optString("access_token").trim()
        if (accessToken.isBlank()) return null

        val refreshToken = json.optString("refresh_token").trim()
        val expiresAt =
            json.optLong("expires_at", -1L).takeIf { it > 0L }
                ?: json.optLong("expires_in", -1L)
                    .takeIf { it > 0L }
                    ?.let { (System.currentTimeMillis() / 1000L) + it }

        val user = json.optJSONObject("user")
        val userId = user?.optString("id")?.trim().orEmpty().ifBlank { null }
        val email = user?.optString("email")?.trim().orEmpty().ifBlank { null }
        val anonymous = user?.optBoolean("is_anonymous", false) ?: false

        return Session(
            accessToken = accessToken,
            refreshToken = refreshToken,
            expiresAtEpochSec = expiresAt,
            userId = userId,
            email = email,
            anonymous = anonymous
        )
    }

    private fun baseHeaders(): Headers {
        return Headers.Builder()
            .add("apikey", supabasePublishableKey)
            .add("Content-Type", "application/json")
            .add("Accept", "application/json")
            .build()
    }

    private fun authHeaders(token: String): Headers {
        return Headers.Builder()
            .addAll(baseHeaders())
            .add("Authorization", "Bearer ${token.trim()}")
            .build()
    }

    private fun requireSuccess(response: CrispyHttpResponse): String {
        if (response.code in 200..299) return response.body
        val message = extractErrorMessage(response.body)
        throw IllegalStateException(message ?: "HTTP ${response.code}")
    }

    private fun extractErrorMessage(rawBody: String): String? {
        val trimmed = rawBody.trim()
        if (trimmed.isBlank()) return null
        return runCatching {
            val parsed = parseJsonBody(trimmed)
            val obj =
                when (parsed) {
                    is JSONObject -> parsed
                    is JSONArray -> parsed.optJSONObject(0)
                    else -> null
                }
            obj?.firstNonBlank(
                "message",
                "msg",
                "error_description",
                "error"
            ) ?: (parsed as? String)?.takeIf { it.isNotBlank() } ?: trimmed
        }.getOrElse { trimmed }
    }

    private fun parseJsonBody(body: String): Any {
        val trimmed = body.trim()
        return when {
            trimmed.startsWith("[") -> JSONArray(trimmed)
            trimmed.startsWith("{") -> JSONObject(trimmed)
            else -> trimmed
        }
    }

    private fun JSONObject.firstNonBlank(vararg keys: String): String? {
        for (key in keys) {
            val value = optString(key).trim()
            if (value.isNotBlank()) return value
        }
        return null
    }

    private fun loadSession(): Session? {
        cachedSession?.let { return it }
        val raw = prefs.getString(KEY_SESSION, null) ?: return null
        val parsed = runCatching {
            val json = JSONObject(raw)
            val accessToken = json.optString("access_token").trim()
            if (accessToken.isBlank()) return null
            val refreshToken = json.optString("refresh_token").trim()
            val expiresAt = json.optLong("expires_at", -1L).takeIf { it > 0L }
            val userId = json.optString("user_id").trim().ifBlank { null }
            val email = json.optString("email").trim().ifBlank { null }
            val anonymous = json.optBoolean("anonymous", false)
            Session(
                accessToken = accessToken,
                refreshToken = refreshToken,
                expiresAtEpochSec = expiresAt,
                userId = userId,
                email = email,
                anonymous = anonymous
            )
        }.getOrNull()
        cachedSession = parsed
        return parsed
    }

    private fun saveSession(session: Session?) {
        cachedSession = session
        if (session == null) {
            prefs.edit().remove(KEY_SESSION).apply()
            return
        }
        val json =
            JSONObject()
                .put("access_token", session.accessToken)
                .put("refresh_token", session.refreshToken)
                .put("expires_at", session.expiresAtEpochSec ?: JSONObject.NULL)
                .put("user_id", session.userId ?: JSONObject.NULL)
                .put("email", session.email ?: JSONObject.NULL)
                .put("anonymous", session.anonymous)
                .toString()

        prefs.edit().putString(KEY_SESSION, json).apply()
    }

    private fun shouldRefresh(session: Session): Boolean {
        val expiresAt = session.expiresAtEpochSec ?: return false
        val nowEpochSec = System.currentTimeMillis() / 1000L
        return expiresAt <= nowEpochSec + SESSION_EXPIRY_SKEW_SEC
    }

    private fun isInvalidRefreshResponse(response: CrispyHttpResponse): Boolean {
        if (response.code !in 400..401) {
            return false
        }
        val message = extractErrorMessage(response.body)?.lowercase().orEmpty()
        if (message.isBlank()) {
            return false
        }
        return message.contains("invalid refresh token") ||
            message.contains("invalid grant") ||
            message.contains("refresh token") && message.contains("invalid") ||
            message.contains("refresh token") && message.contains("expired") ||
            message.contains("session_not_found")
    }

    private sealed interface RefreshResult {
        data class Success(val session: Session) : RefreshResult

        data class TransientFailure(val message: String?) : RefreshResult

        data object InvalidSession : RefreshResult
    }

    private companion object {
        private const val PREFS_NAME = "supabase_sync_lab"
        private const val KEY_SESSION = "@supabase:session"

        private const val CALL_TIMEOUT_MS = 10_000L
        private const val SESSION_EXPIRY_SKEW_SEC = 30L
    }
}

File: ios/CrispyKit/Sources/CrispyKit/Auth/SupabaseAccountClient.swift
import Foundation

/// URLSession port of the Android `SupabaseAccountClient` (email sign-in/up,
/// token refresh, sign-out). Envelope handling matches the Kotlin client.
public final class SupabaseAccountClient {
    public struct SignUpResult {
        public let session: Session?
        public let message: String
    }

    private let httpClient: CrispyHttpClient
    private let baseURL: String
    private let publishableKey: String
    private let tokenStore: SessionStoring
    private let nowEpochSeconds: () -> Int64

    private var refreshInFlight: Task<Session?, Never>?

public init(
        httpClient: CrispyHttpClient,
        supabaseURL: String,
        publishableKey: String,
        tokenStore: SessionStoring,
        nowEpochSeconds: @escaping () -> Int64 = { Int64(Date().timeIntervalSince1970) }
    ) {
        self.httpClient = httpClient
        self.baseURL = supabaseURL.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        self.publishableKey = publishableKey.trimmingCharacters(in: .whitespacesAndNewlines)
        self.tokenStore = tokenStore
        self.nowEpochSeconds = nowEpochSeconds
    }

public func isConfigured() -> Bool {
        !baseURL.isEmpty && !publishableKey.isEmpty
    }

public func currentSession() -> Session? {
        tokenStore.current()
    }

public func ensureValidSession() async -> Session? {
        guard let existing = tokenStore.current() else { return nil }
        if !shouldRefresh(existing) { return existing }
        if !isConfigured() {
            saveSession(nil)
            return nil
        }

        if let inFlight = refreshInFlight {
            return await inFlight.value
        }
        let task = Task<Session?, Never> { [weak self] in
            guard let self else { return nil }
            return await self.refreshLocked()
        }
        refreshInFlight = task
        defer { refreshInFlight = nil }
        return await task.value
    }

public func signInWithEmail(email: String, password: String) async throws -> Session {
        try checkConfigured()
        let payload: [String: Any] = [
            "email": email.trimmingCharacters(in: .whitespacesAndNewlines),
            "password": password,
        ]
        let body = try await postAuth(path: "/auth/v1/token?grant_type=password", payload: payload)
        guard let session = parseSession(try JsonParser.parseObject(body)) else {
            throw AuthError.signinFailed("Sign-in did not return a session.")
        }
        saveSession(session)
        return session
    }

public func signUpWithEmail(email: String, password: String) async throws -> SignUpResult {
        try checkConfigured()
        let payload: [String: Any] = [
            "email": email.trimmingCharacters(in: .whitespacesAndNewlines),
            "password": password,
        ]
        let body = try await postAuth(path: "/auth/v1/signup", payload: payload)
        let json = try JsonParser.parseObject(body)
        if let session = parseSession(json) {
            saveSession(session)
            return SignUpResult(session: session, message: "Account created and signed in.")
        }
        let hasUser = json.jsonObject("user")?.jsonString("id") != nil
        return SignUpResult(
            session: nil,
            message: hasUser ? "Account created. Confirm your email, then sign in." : "Account created."
        )
    }

    /// Revokes the server-side session. Local wipe stays with the caller.
public func signOut() async {
        guard isConfigured(), let session = tokenStore.current() else { return }
        guard !session.accessToken.hasPrefix("cp_pat_") else { return }
        _ = try? await postAuth(
            path: "/auth/v1/logout",
            payload: ["scope": "global"],
            bearerToken: session.accessToken
        )
    }

    // MARK: - Refresh

    private func refreshLocked() async -> Session? {
        guard let latest = tokenStore.current() else { return nil }
        if !shouldRefresh(latest) { return latest }
        guard !latest.refreshToken.isEmpty else {
            saveSession(nil)
            return nil
        }

        do {
            let body = try await postAuth(
                path: "/auth/v1/token?grant_type=refresh_token",
                payload: ["refresh_token": latest.refreshToken]
            )
            guard let session = parseSession(try JsonParser.parseObject(body)) else {
                return latest
            }
            saveSession(session)
            return session
        } catch let error as AuthError {
            if case .invalidRefreshToken = error {
                saveSession(nil)
                return nil
            }
            return latest
        } catch {
            return latest
        }
    }

    // MARK: - Plumbing

    private func parseSession(_ json: [String: Any]) -> Session? {
        guard let accessToken = json.jsonString("access_token") else { return nil }
        var expiresAt = json.jsonInt("expires_at").map(Int64.init).flatMap { $0 > 0 ? $0 : nil }
        if expiresAt == nil, let expiresIn = json.jsonInt("expires_in").map(Int64.init), expiresIn > 0 {
            expiresAt = nowEpochSeconds() + expiresIn
        }
        let user = json.jsonObject("user")
        return Session(
            accessToken: accessToken,
            refreshToken: json.jsonString("refresh_token") ?? "",
            expiresAtEpochSec: expiresAt,
            userId: user?.jsonString("id"),
            email: user?.jsonString("email"),
            anonymous: user?.jsonBool("is_anonymous", defaultValue: false) ?? false
        )
    }

    private func shouldRefresh(_ session: Session) -> Bool {
        guard let expiresAt = session.expiresAtEpochSec else { return false }
        return expiresAt <= nowEpochSeconds() + Self.expirySkewSeconds
    }

    private func baseHeaders() -> [String: String] {
        [
            "apikey": publishableKey,
            "Content-Type": "application/json",
            "Accept": "application/json",
        ]
    }

    private func authHeaders(_ token: String) -> [String: String] {
        var headers = baseHeaders()
        headers["Authorization"] = "Bearer \(token.trimmingCharacters(in: .whitespacesAndNewlines))"
        return headers
    }

    private func requireConfiguredBaseURL() throws -> String {
        try checkConfigured()
        return baseURL
    }

    private func checkConfigured() throws {
        if !isConfigured() {
            throw AuthError.notConfigured
        }
    }

    private func postAuth(path: String, payload: [String: Any], bearerToken: String? = nil) async throws -> String {
        let url = URL(string: try requireConfiguredBaseURL() + path)
        let headers = bearerToken.map(authHeaders) ?? baseHeaders()
        let response = try await httpClient.postJson(
            url: try requireValid(url),
            jsonBody: try JsonParser.encodeObject(payload),
            headers: headers,
            timeout: Self.callTimeoutSeconds
        )
        guard (200...299).contains(response.code) else {
            throw authError(code: response.code, body: response.body)
        }
        return response.body
    }

    private func requireValid(_ url: URL?) throws -> URL {
        guard let url else { throw AuthError.notConfigured }
        return url
    }

    private func authError(code: Int, body: String) -> AuthError {
        let message = extractErrorMessage(body)
        if (400...401).contains(code), let lowered = message?.lowercased() {
            let invalidRefresh =
                lowered.contains("invalid refresh token") ||
                lowered.contains("invalid grant") ||
                (lowered.contains("refresh token") && lowered.contains("invalid")) ||
                (lowered.contains("refresh token") && lowered.contains("expired")) ||
                lowered.contains("session_not_found")
            if invalidRefresh {
                return .invalidRefreshToken
            }
        }
        return .http(code: code, message: message ?? "HTTP \(code)")
    }

    private func extractErrorMessage(_ rawBody: String) -> String? {
        let trimmed = rawBody.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        guard let parsed = try? JsonParser.parseObject(trimmed) else { return trimmed }
        for key in ["message", "msg", "error_description", "error"] {
            if let value = parsed.jsonString(key) { return value }
        }
        return trimmed
    }

    private func saveSession(_ session: Session?) {
        if let session {
            tokenStore.save(session)
        } else {
            tokenStore.clear()
        }
    }

    public enum AuthError: Error {
        case notConfigured
        case signinFailed(String)
        case invalidRefreshToken
        case http(code: Int, message: String)

        public var localizedMessage: String {
            switch self {
            case .notConfigured:
                return "Supabase is not configured."
            case .signinFailed(let message):
                return message
            case .invalidRefreshToken:
                return "Session expired. Please sign in again."
            case .http(_, let message):
                return message
            }
        }
    }

    private static let callTimeoutSeconds: TimeInterval = 10
    private static let expirySkewSeconds: Int64 = 30
}
